Firehose

Eduard-Voiculescu opened the pull request:
<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺
v                               ✰  Thanks for creating a PR! ✰    
v    Before smashing the submit button please review the checkboxes.
v    If a checkbox is n/a - please still include it but + a little note why
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  -->

Closes: #XXX

## Description

<!-- Add a description of the changes that this PR introduces and the files that
are the most critical to review.
-->

______

For contributor use:

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added  relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer

______

For admin use:

- [ ] Added appropriate labels to PR (ex. `WIP`, `R4R`, `docs`, etc)
- [ ] Reviewers assigned
- [ ] Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Release Notes

- **New Features**
	- Introduced enhanced tracing capabilities for Ethereum transactions using the Firehose protocol.
	- Added new tracing hooks for block and transaction events, improving observability.
	- Expanded the `.gitignore` to include environment files.
	- Implemented new functions for managing Ethereum block events within the Cosmos SDK.
	- Updated EVM configuration to support additional tracer types, including "firehose."
	- Added a new error handling mechanism for state overrides.

- **Bug Fixes**
	- Improved error handling for invalid trace configurations in tests.

- **Documentation**
	- Updated configuration templates to include new tracer types.

- **Tests**
	- Added new test cases to validate tracing functionalities and enhance coverage for state transitions and balance changes.

- **Chores**
	- Updated dependency versions for improved compatibility and performance.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

arrivets committed:
feat: add firehose tracer

arrivets committed:
chore: add firehose test case in TestApplyMessageWithConfig

arrivets committed:
chore: fix firehose test case

arrivets committed:
chore: make firehose the default tracer

This breaks some tests and shows that we need to add calls to OnBlockStart to
prevent panics.

arrivets committed:
chore: cleanup comment about default tracer

arrivets committed:
chore: move test case with firehose tracer

Eduard-Voiculescu committed:
- add firehose evm tracer hooks
- move some files around to keep them contained inside of the x/evm package
- adding env file ignore in gitignore

Eduard-Voiculescu committed:
Tracer refactoring

- Removing tracer from the EthermintApp struct
- Updating tests which referenced the tracer from the EVMKeeper
- Adding InitChainer method on keeper to call when a block chain is initialized

Eduard-Voiculescu committed:
Adding EndBlock tracer and using the k.evmTracer instead of the tracer from the config

Eduard-Voiculescu committed:
Update finalized header when begin block hook is called

Eduard-Voiculescu committed:
Removing unused method and setting header values

arrivets commented on app/app.go:
```diff
@@ -1112,3 +1112,43 @@ func initParamsKeeper(appCodec codec.BinaryCodec, legacyAmino *codec.LegacyAmino
 }
+
+// Call this method in to OnBlockStart call
+func TmBlockHeaderToEVM(
```
Should this be moved to evm/keeper/abci.go ?
Looks like it's not being used yet

Eduard-Voiculescu replied:
Yep, I think that we can actually remove this

Eduard-Voiculescu replied:
It has been removed

arrivets commented on x/evm/types/tracer.go:
```diff
@@ -55,7 +60,10 @@
 	case TracerStruct:
 		hooks = logger.NewStructLogger(logCfg).Hooks()
+	case Firehose:
+		hooks, _ = tracers.LiveDirectory.New("firehose", nil)
```
Do we need to pass some configuration when instantiating the firehose tracer?

arrivets commented on x/evm/types/tracer.go:
```diff
@@ -64,6 +72,17 @@ func NewTracer(tracer string, msg *core.Message, rules params.Rules) *tracers.Tr
 	}
 }
 
+func NewLiveTracer(tracer string) (*tracers.Tracer, error) {
```
This isn't used

Eduard-Voiculescu replied:
Yep, good point, removing it

arrivets commented on x/evm/keeper/state_transition_test.go:
```diff
@@ -671,13 +671,11 @@ func (suite *StateTransitionTestSuite) TestApplyMessageWithConfig() {
 				)
 				suite.Require().NoError(err)
-
-				config.Tracer = types.NewTracer("firehose", msg, params.Rules{})
```
Without this line we are not actually testing the firehose tracer

Eduard-Voiculescu replied:
I created a test called `TestApplyMessageWithConfigTracer` as I do a bunch of different stuff there

Eduard-Voiculescu committed:
Update tracer with custom OnCosmosBlockStart and OnCosmosBlockEnd hooks

Eduard-Voiculescu committed:
fix block header differences bet/ firehose tracer and eth_getBlock method call

Eduard-Voiculescu committed:
removing debug log

Eduard-Voiculescu committed:
update statedb tracer

- add tests for statedb balance change
- fix add and sub balance changes
- update dependencies to cometbft and cosmos-sdk

Eduard-Voiculescu committed:
adding balance change setBalance changes

Eduard-Voiculescu committed:
Update tracer tests for nonce changes, code changes and storage changes

Eduard-Voiculescu committed:
Bumping cometbft and cosmos-sdk versions

Eduard-Voiculescu committed:
Update cometbft dependency with sf fork with changes done to PartSetHeader

Eduard-Voiculescu committed:
Update tracer

BeginBlock and EndBlock will fetch the tracer from the sdk.Context, when calling ApplyTransaction, we will fetch the tracer from the keeper.evmTracer

Eduard-Voiculescu committed:
Fix on tx end hook call

Eduard-Voiculescu committed:
Fix outdated grpc_query tests

Eduard-Voiculescu committed:
Adding firehose trace in the OnStorageChange call

Eduard-Voiculescu committed:
Adding test for ApplyTransactionWithTracer to check that the hooks are properly called

Eduard-Voiculescu commented on x/evm/tracers/firehose_test.go:
```diff
@@ -0,0 +1,493 @@
+			f := NewFirehose(&FirehoseConfig{
+				ApplyBackwardCompatibility: ptr(false),
+			})
+			f.OnBlockchainInit(params.AllEthashProtocolChanges)
```
ptr function is using generics, so the function will return the type `*type` of what is passed in as the first argument

Eduard-Voiculescu commented on x/evm/keeper/state_transition.go:
```diff
@@ -312,7 +314,7 @@ func (k *Keeper) ApplyMessageWithConfig(
 	cfg *EVMConfig,
 	commit bool,
-) (*types.MsgEthereumTxResponse, error) {
+) (resp *types.MsgEthereumTxResponse, err error) {
```
All good

Eduard-Voiculescu committed:
- address pr comments
- move tracer calls to storage changes higher up

Eduard-Voiculescu committed:
Fix issue in tracer not catching correct error in `state_transition.go`

update with suggestions from coderabbitai

Eduard-Voiculescu commented on x/evm/handler_test.go:
```diff
@@ -104,7 +104,7 @@ func (suite *HandlerTestSuite) TestHandleMsgEthereumTx() {
 		},
 		{
-			"insufficient balance",
+			"insufficient balance", // todo: shouldn't this be insufficient gas
 			func() {
 				tx = types.NewTxContract(suite.chainID, 0, big.NewInt(100), 0, big.NewInt(10000), nil, nil, nil, nil)
 				suite.signTx(tx)
```
Removed

Eduard-Voiculescu commented on x/evm/tracers/firehose.go:
```diff
@@ -0,0 +1,2436 @@
+	f.block.Header.Hash = f.cosmosBlockHeader.Hash()
+	f.block.Hash = f.block.Header.Hash
+
+	// todo: find the right size
+	f.block.Size = 10
```
Resolving this for the moment, but we need to find a way to fetch the correct block size.

Eduard-Voiculescu committed:
removing todo

Eduard-Voiculescu committed:
removing todo comment

Eduard-Voiculescu commented on x/evm/statedb/statedb.go:
```diff
@@ -423,83 +438,154 @@ func (s *StateDB) Transfer(sender, recipient common.Address, amount *big.Int) {
 		s.err = err
 	}
+
+	if s.err == nil {
```
@arrivets Actually not sure about this one. Would we want to trace event if the native action has returned an error?

arrivets replied:
No I don't think so. Seems correct to only trace if there is no error

Eduard-Voiculescu commented on x/evm/statedb/statedb.go:
```diff
@@ -423,83 +438,154 @@ func (s *StateDB) Transfer(sender, recipient common.Address, amount *big.Int) {
 		s.err = err
 	}
+
+	if s.err == nil {
```
Same point here

Eduard-Voiculescu commented on x/evm/statedb/statedb.go:
```diff
@@ -423,83 +438,154 @@ func (s *StateDB) Transfer(sender, recipient common.Address, amount *big.Int) {
 		s.err = err
 	}
+
+	if s.err == nil {
```
Same here

Eduard-Voiculescu commented on x/evm/statedb/statedb.go:
```diff
@@ -423,83 +438,154 @@ func (s *StateDB) Transfer(sender, recipient common.Address, amount *big.Int) {
 		s.err = err
 	}
+
+	if s.err == nil {
```
Same here

Eduard-Voiculescu committed:
Removing global registry

arrivets committed:
refactor: removce EVMConfigWithTracer

arrivets committed:
refactor: comment choice of tracer in prepareTrace

arrivets committed:
refactor: update comment on NewEVM about RANDAO

arrivets committed:
chore: add nil check in StateDB.AddLog

arrivets committed:
chore: remove duplicate package alias in statedb_test.go

arrivets committed:
refactor: delete unused LiveTracerRegistry

arrivets committed:
refactor: delete unused BlockchainTraceFactory

arrivets committed:
refactor: change name of methods that get/set tracing hooks in context

arrivets committed:
refactor: fix imports in x/evm/types/tracer.go

arrivets committed:
refactor: remove unused NewLiveTracer

arrivets commented on server/config/toml.go:
```diff
@@ -25,7 +25,7 @@ const DefaultConfigTemplate = `
 # Tracer defines the 'vm.Tracer' type that the EVM will use when the node is run in
 # debug mode. To enable tracing use the '--evm.tracer' flag when starting your node.
-# Valid types are: json|struct|access_list|markdown
+# Valid types are: json|struct|access_list|markdown|firehose
```
Hi @Eduard-Voiculescu 
Could you add something in app.go that uses this config to set the required tracer?

arrivets reviewed (commented):
Looks good overall.

It would be good to add something in app.go that reads the `evm.tracer` config flag and sets the tracer appropriately. For all the possible values of that flag

Eduard-Voiculescu committed:
removing the firehose switch case from the NewTracer method as we are using custom hooks for the firehose

arrivets committed:
chore: add tracer in app.go

arrivets commented on x/evm/types/tracer.go:
```diff
@@ -64,6 +70,15 @@ func NewTracer(tracer string, msg *core.Message, rules params.Rules) *tracers.Tr
 	}
 }
 
+func NewFirehoseCosmosLiveTracer() (*cosmostracing.Hooks, error) {
+	h, err := cosmostracers.NewCosmosFirehoseTracer(false)
+	if err != nil {
+		return nil, fmt.Errorf("initializing live tracer firehose: %w", err)
+	}
+
+	return h, nil
+}
```
not possible because Firehose doesn't use `tracing.Hooks` (from go-ethereum/core/tracing), but custom hooks (from x/evm/tracing)

arrivets committed:
chore: refactor initializeEVM

arrivets committed:
chore: define error type for ErrConfigOverrides

arrivets committed:
chore: update TestApplyMessage

arrivets committed:
chore: remove manual call to OnNonceChange after evm.Call

xlab commented on x/evm/statedb/statedb.go:
```diff
@@ -423,83 +439,154 @@ func (s *StateDB) Transfer(sender, recipient common.Address, amount *big.Int) {
 		panic("negative amount")
 	}
+
+	balance := s.GetBalance(addr)
```
collect only if `s.evmTracer != nil && s.evmTracer.OnBalanceChange != nil`, reduce reads to StateDB if tracer is not active

Eduard-Voiculescu replied:
Same point here as mentioned below, Martin has moved it inside of the check if the tracer is set

xlab commented on x/evm/statedb/statedb.go:
```diff
@@ -423,83 +439,154 @@ func (s *StateDB) Transfer(sender, recipient common.Address, amount *big.Int) {
 
 	coins := sdk.NewCoins(sdk.NewCoin(s.evmDenom, sdkmath.NewIntFromBigIntMut(amount)))
 	senderAddr := sdk.AccAddress(sender.Bytes())
+	senderBalance := s.GetBalance(sender)
```
Collect these only if `s.evmTracer != nil && s.evmTracer.OnBalanceChange != nil`, reduce reads to StateDB if tracer is not active

Eduard-Voiculescu replied:
I see that martin has made the changes for this

kakysha commented on x/evm/keeper/config.go:
```diff
@@ -149,9 +153,14 @@ func (k Keeper) VMConfig(ctx sdk.Context, cfg *EVMConfig) vm.Config {
 
-	return vm.Config{
-		Tracer:    cfg.Tracer.Hooks,
+	vmCfg := vm.Config{
```
all of these changes has no effect, they are similar to what it was before

Eduard-Voiculescu replied:
Maybe, but I think it's still safer to not just assign `cfg.Tracer.Hooks` to `vmConfig.Tracer` as `cfg.Tracer.Hooks` could unmistakably be nil. We could have a situation where the tracer is not properly configured in the future if there is a change in the code and someone forgets to add a nil check here.
